src/transports/shared.rs

```diff
@@ -0,0 +1,139 @@
+use std::{thread, fmt, mem};
+use std::sync::{self, atomic, Arc};
+use futures::{self, Future};
+use futures::sync::oneshot;
+use transports::Result;
+use transports::tokio_core::reactor;
+use {Error, RequestId};
+
+/// Event Loop Handle.
+/// NOTE: Event loop is stopped when handle is dropped!
+pub struct EventLoopHandle {
+  thread: Option<thread::JoinHandle<()>>,
+  remote: reactor::Remote,
+  done: Arc<atomic::AtomicBool>,
+}
+
+impl EventLoopHandle {
+  /// Creates a new `EventLoopHandle` and transport given the transport initializer.
+  pub fn spawn<T, F>(func: F) -> Result<(Self, T)> where
+    F: FnOnce(&reactor::Handle) -> Result<T>,
+    F: Send + 'static,
+    T: Send + 'static,
+  {
+    let done = Arc::new(atomic::AtomicBool::new(false));
+    let (tx, rx) = sync::mpsc::sync_channel(1);
+    let done2 = done.clone();
+
+    let eloop = thread::spawn(move || {
+      let run = move || {
+        let event_loop = reactor::Core::new()?;
+        let http = func(&event_loop.handle())?;
+        Ok((http, event_loop))
+      };
+
+      let send = move |result| {
+        tx.send(result).expect("Receiving end is always waiting.");
+      };
+
+      let res = run();
+      match res {
+        Err(e) => send(Err(e)),
+        Ok((http, mut event_loop)) => {
+          send(Ok((http, event_loop.remote())));
+
+          while !done2.load(atomic::Ordering::Relaxed) {
+            event_loop.turn(None);
```

Comment:
but that isn't what we want, is it? We would prefer for the thread to join when the flag is set.